src/window/city.c

```diff
@@ -155,6 +161,100 @@ static void cycle_legion(void)
     }
 }
 
+/** 
```

Comment:
All these added functions is rather a lot of logic to put in `window/city.c`, which should only be concerned with handling input events.
I suggest creating a new C file `building/clone.c` that contains this logic, with one public function `building_clone_from_grid_offset(int grid_offset)`, which has the implementation of `get_clone_type_from_grid_offset()`.